"compare" and "merge" helper scripts overhaul

MCStreetguy opened the pull request:
Like we've previously discussed, I have rewritten the "compare" and "merge" helper scripts entirely in Node JS and extended their functionality quite a bit, in the hopes of simplifying the translation process and making it more error-proof. I tested the new scripts already against the D&D system translation files of the current and upcoming version and everything seems to be working well. That being said, I will mark the pull request as a draft for now, as I definitely want to test it more rigorously before merging, but I'd love hear your early opinion on my proposal and if there is anything you'd like me to add or change.

As the complexity of the scripts pretty much exploded compared to their python predecessors, I'll explain their functionality and my reasoning below.

### Comparison process

First and foremost, the comparison process has been extended to account for more than just missing keys between the `de.json` and `en.json` files. The following cases are now covered:
- Keys that are missing entirely from the `de.json` translation file
- Keys that have a different type between the two files, e.g. simple strings that have become objects or maybe also the other way around
- Keys that have been removed from the `en.json` translation file and thus might be obsolete  
_(this is optional and only checked when the option `--removed` is passed to the script)_
- Keys that are identical between the two files, to find possible errornously merged but still untranslated labels in `de.json`  
_(also optional and only enabled with the `--duplicates` option)_

All of these checks are done recursively whenever an object is encountered, so all nested keys at any depth are checked too.

The last two checks are disabled by default, because these require more thorough manual validation of the results. Especially the duplicates check, as there are a lot of cases in which it is totally fine to have the same label in both languages and this extra workload might not always be desired. For removed keys on the other hand, we might not always want to get rid of them immediately to retain compatibility with previous versions.

This is also the reason I didn't explain these additional options in the README and only mentioned the missing and updated steps to keep the process as easy as possible for new contributors that might only want to add a few missing translations and not bother with these "advanced" workflows.

### General changes

Instead of having a single intermediate diff file, the `compare.mjs` script will now generate up to four separate files for each of the cases listed above. This is meant to make it easier to differentiate how to handle the respective keys in them.

The intermediate translation diffs now also use the JSON format themselves, in contrast to the custom txt formatting the python scripts used. This ensures that formatting errors _(such as the one that just happened in my last contribution)_ should not occur anymore - if it's valid JSON in the intermediate step, it will also be valid once merged back into `de.json`, and if it isn't, well, it simply cannot be merged anyway. Also it was much simpler just relying on `JSON.parse` and `JSON.stringify` instead of re-creating the custom txt encoding, to be honest.

As a result, with all comparison options enabled (and if there are actually matches for each of them) we would get four diff files in the `languages` directory:
- `_missing.json`
- `_updated.json`
- `_removed.json`
- `_duplicates.json`

Also I've moved the new scripts into a new folder `util/` to prevent cluttering the root directory of the project and possibly confusing the development helpers with the actual module logic.

### Merging process

The merge script isn't actually behaving that different to before, other than it working with up to four input files and also recursively merging all translations.

I've kept the behaviour that it won't overwrite existing keys in `de.json` by default, which can optionally enabled by passing it the `--overwrite` option. This replaces the previously hardcoded flag from the python script. Without the option present, it will skip the `_updated.json` and `_duplicates.json` files entirely, as these require overwriting existing keys in any case. In addition, if for some unknown reason there is an already existing key in the `_missing.json` file, that would also not be applied.

The `sort_only` flag is no longer supported (even though it wasn't used in the python script either as far as i can tell), instead the script will now always sort the resulting object to hopefully prevent extensive git diffs by enforcing a fixed order for all keys. This will also happen if there wasn't actually anything to merge into the `de.json` file so executing the script without any of the four possible intermediate diff files will actually behave like I imagine the `sort_only` flag was once intended.

Finally, the merge script will not remove any keys by default, regardless of wether the `_removed.json` file exists and contains any keys to delete. This will only happen if the `--removed` option has been passed to it, which is intended as an additional safeguard against accidental removal of translations, similar to the `--overwrite` option.

### Misc / Logging

If either the `de.json` or `en.json` files cannot be found, the scripts will abort with an error message. In the latter case the URL of the latest `en.json` from the D&D system will also be appended together with a note to please download it:

```
$ node utils/compare.mjs
ERROR: Cannot find "en.json" file in "/home/.../foundryvtt-dnd5e-lang-de/languages"! Please download the latest version of the file from https://github.com/foundryvtt/dnd5e/raw/master/lang/en.json and place it in the "languages" directory.
```

Both scripts will output a summary at the end:

```
$ node utils/compare.mjs --duplicates --removed
104 keys are identical in both translation files. You can find them in the "_duplicates.json" file.
13 keys are missing a translation in de.json. You can find them in the "_missing.json" file.
1 keys have an updated type in en.json. You can find them in the "_updated.json" file.
31 keys are missing from en.json. You can find them in the "_removed.json" file.

$ node utils/merge.mjs --overwrite --removed
Successfully updated de.json.
13 translations have been added.
68 translations have been updated.
31 translations have been removed.
```

_(yes, that is indeed the result when comparing the current `de.json` against the latest `en.json` and not just a demo. I'll take care of it as soon as the scripts are finished and verified to be working flawlessly. I've also checked the reported duplicates and there are surprisingly at least 7 actually untranslated keys in the current `de.json` file. Also, I'm aware of the discrepancy between the amount of changes both scripts report and already looking into why there are less updates than detected differences, when normally these numbers should match as I've just merged the untouched diff JSONs.)_

I'm looking forward to your feedback to my proposal. If there is anything still unclear please let me know.
In case you want to try the new scripts out for yourself you can get them [from my fork](https://github.com/MCStreetguy/foundryvtt-dnd5e-lang-de/tree/master/utils).

MCStreetguy commented:
I finally found the problem when merging objects. In fact, the numbers were correct, the error was that I had forgotten to pass the “overwrite” parameter to the recursive call in the merge function. Classic...

Now both the displayed numbers and the merged translations in the `de.json` are correct. I checked this again manually for each individual key and also ran a script over it. Better safe than sorry, and I also wanted to make sure that I didn't overlook something again, let alone incorporate such a bug into the new scripts...

The output now looks like this:

```
$ node utils/compare.mjs --duplicates --removed
104 keys are identical in both translation files. You can find them in the "_duplicates.json" file.
13 keys are missing a translation in de.json. You can find them in the "_missing.json" file.
1 keys have an updated type in en.json. You can find them in the "_updated.json" file.
31 keys are missing from en.json. You can find them in the "_removed.json" file.

$ node utils/merge.mjs --overwrite --removed
Successfully updated de.json.
13 translations have been added.
105 translations have been updated.
31 translations have been removed.
```

From my side, I would call the rework of the scripts completed, so I will mark the pull request as ready for review. Of course, I'm still looking forward to your feedback on the changes.

MCStreetguy commented:
I also had a look at the other helper scripts in the repo to see if I could update them as well. Unfortunately I don't really know what to do with either the `compile_csv.py` or the `spellcheck.py` script. With the former I'm not sure what exactly it does and with the latter I had the problem that I'm not familiar with LanguageTool and would have to familiarize myself with it first.

But I actually noticed something with the third one. I'm afraid the `packs_to_babele.py` script has become obsolete. From what it looks like to me, it was written for the old NeDB format that Foundry used before v11. Now the packs are all stored in LevelDB format, which is no longer easily readable as a text file. As it happens, I've only recently been looking into how to deal with the format, because I wanted to version my own compendiums and prefer to do that in JSON format like the D&D system instead of having to upload binary files to Git.

The official Foundry CLI available via NPM exposes two functions for compiling and unpacking compendia from/to JSON or YAML _(see <https://www.npmjs.com/package/@foundryvtt/foundryvtt-cli#api>)_. That is also how the D&D system deals with LevelDB for Git _(see <https://github.com/foundryvtt/dnd5e/blob/4.2.x/utils/packs.mjs>)_. I believe these functions could certainly be used to achieve similar functionality, although I think we would have to do significant transformations to the unpacked JSON files to match the format Babele expects for translation files.

I still need to learn more about this, but wanted to point out that we would probably need a complete rework for `packs_to_babele.py` or alternatively do the compendia translations by hand.

mhilbrunner commented:
This is looking super exciting! Thanks so much for working on it! I'll try to find a large block of time to test and review soon (feel free to ping and remind me if you haven't seen a sign of life for longer than a week :P).

Already, the improvements are looking awesome at first glance, you have gone way further than just a translation!

> I'm afraid the packs_to_babele.py script has become obsolete.

Definitely, and don't try to translate that one, as you say, the official CLI is a good replacement for now. This is all super old stuff and pre-dates the LevelDB change; *old* like Foundry 0.7 or so. All the compendium translation workflow needs a complete rework. In which case, we should also probably finally move this to a Weblate-based workflow so it is easier to contribute for non-devs.

(I stopped doing much besides maintenance on this module recently, both due to being quite busy at the moment and with the 5E module becoming *official*, I had (somewhat) hoped this may become obsolete as an official translation may be coming, but so far no news apparently, so 🤷)

MCStreetguy commented:
You're very welcome and thanks again for your kind words! Actually had a lot of fun tinkering with these. Originally I only wanted to add a check to see if the data types of the translations have changed, but then one thing led to another and somehow I ended up including all types of changes I could think of. I thought, can only benefit us, so why not include everything directly while I'm at it anyway. :)
Take your time to review, no pressure please, and hit me up anytime if somethings still unclear.

Moving the compendium workflow to Weblate would certainly make the translations a whole lot easier. I've also had a few thoughts on how best to approach the export of compendia to Babele JSON files, which would be directly compatible with Weblate afaik, but before my comment here gets too long, I'd say we can discuss this in a dedicated issue for the topic.

Yeah, sadly it doesn't look like there will be official german translations any time soon. I hope the maintenance of the module is not too much trouble for you as long as there are no official translations. I also have a lot going on most of the time _(still waiting for the "infinite free time patch" haha)_ but I will definitely help wherever I can!

MCStreetguy commented:
Just added another little feature to the new "merge" helper script, which automatically generates a summary of the changed keys in markdown formatting to be used in corresponding PRs. Also added a note in the README translation guide to please include that summary in the PR description as well as a notice to remove all keys from the diff json's which are still untranslated before merging the changes back into `de.json` to hopefully reduce the likelyhood of untranslated strings slipping through.

The output of `merge.mjs` now looks like this:

```
$ ./utils/merge.mjs --overwrite --removed
Successfully updated de.json.
8 translations have been added.
21 translations have been updated.
31 translations have been removed.

Markdown summary for PR:
------------------------
## Summary of changes

### Added

- `DOCUMENT.DND5E > Warning > SelectType`
- `DND5E.ACTIVITY > Warning > UsageNotAllowed`
- `DND5E.ATTACK > Attack > mwak`
- `DND5E.ATTACK > Attack > rwak`
- `DND5E.ATTACK > Weapon > Melee`
- `DND5E.ATTACK > Weapon > Ranged`
- `DND5E.CONSUMPTION > Warning > WillDestroy`
- `DND5E.CurrencyManager > Error > InsufficientFunds`
- `DND5E.Feature > SpeciesLegacy`
- `DND5E.SUMMON > FIELDS > match > ability > label`
- `DND5E.SUMMON > FIELDS > match > ability > hint`

### Updated

- `DND5E.AreaOfEffect > Label`
- `DND5E.AreaOfEffect > Size > Label`
- `DND5E.AreaOfEffect > Size > Height`
- `DND5E.AreaOfEffect > Size > Length`
- `DND5E.AreaOfEffect > Size > Radius`
- `DND5E.AreaOfEffect > Size > Thickness`
- `DND5E.AreaOfEffect > Size > Width`
- `DND5E.AbilityUseConsumableLabel`
- `DND5E.AdvancementTraitModeUpgradeLabel`
- `DND5E.Award > Message`
- `DND5E.ConsumeAttribute`
- `DND5E.ConsumeScalingLabel`
- `DND5E.Flat`
- `DND5E.FormulaMalformedError`
- `DND5E.FormulaMissingReferenceWarn`
- `DND5E.Issues`
- `DND5E.TARGET > FIELDS > target > template > height > hint`
- `DND5E.UseItem`
- `DND5E.per`
- `JOURNALENTRYPAGE.DND5E > Subclass > FIELDS > style > hint`
- `SOURCE.BOOK.SRD`

### Removed

- `DND5E.AbilityUseChargedHint`
- `DND5E.AbilityUseConsumableChargeHint`
- `DND5E.AbilityUseConsumableDestroyHint`
- `DND5E.AbilityUseConsumableDestroyResourceHint`
- `DND5E.AbilityUseConsumableQuantityHint`
- `DND5E.AbilityUseConsume`
- `DND5E.AbilityUseConsumeAction`
- `DND5E.AbilityUseHint`
- `DND5E.AbilityUseNormalHint`
- `DND5E.AbilityUseRechargeHint`
- `DND5E.AbilityUseUnavailableHint`
- `DND5E.EquipmentClothing`
- `DND5E.EquipmentTrinket`
- `DND5E.EquipmentVehicle`
- `DND5E.Summoning > Action`
- `DND5E.Summoning > Bonuses`
- `DND5E.Summoning > Configuration`
- `DND5E.Summoning > Count`
- `DND5E.Summoning > CreatureChanges`
- `DND5E.Summoning > CreatureSizes`
- `DND5E.Summoning > CreatureTypes`
- `DND5E.Summoning > DisplayName`
- `DND5E.Summoning > DropHint`
- `DND5E.Summoning > ItemChanges`
- `DND5E.Summoning > Level`
- `DND5E.Summoning > Match`
- `DND5E.Summoning > Mode`
- `DND5E.Summoning > Profile`
- `DND5E.Summoning > Prompt`
- `DND5E.Summoning > Warning`
- `DND5E.WeaponOtherProficiency`

```

There may be a slight difference between the number of changes and the summary lists due to how the script handles this information. The number of changes is based on how many keys the `deepMergeObjects` function actually changed. So if it overwrites a string with an object with, say, 7 keys (as in the `DND5E.AreaOfEffect` case above), it's actually counted as only a single change (the object instead of the previous string). But in the summary list, all 7 contained keys of that object are listed individually, hence the discrepancy.

mhilbrunner commented:
This module is being sunset, as discussed in https://github.com/mhilbrunner/foundryvtt-dnd5e-lang-de/issues/41#issuecomment-4296819410.
